resources: update gapbs to use new 24.04 images (#46)
gem5/gem5-resources: src/gapbs/README.md

permalink: resources/gapbs
shortdoc: >
    This resource implementes the [GAP benchmark suite](http://gap.cs.berkeley.edu/benchmark.html).
author: ["Harshil Patel"]
license: BSD-3-Clause
---

This document provides instructions to create a GAP Benchmark Suite (GAPBS) disk image, which, along with an example script, may be used to run GAPBS within gem5 simulations. The example script uses a pre-built disk-image.

> A pre-built disk image, for X86, can be found, gzipped, here: <add link>

Reviewer: Add link

> A pre-built disk image, for X86, can be found, gzipped, here: <add link>

Reviewer: Add link

Author: If everything looks good then I will upload the disks and add the links.

## What's on the disk?

- username: gem5
- password: 12345

- The `gem5-bridge`(m5) utility is installed in `/usr/local/bin/gem5-bridge`.
- `libm5` is installed in `/usr/local/lib/`.

> - The headers for `libm5` are installed in `/usr/local/include/gem5-bridge`.

Reviewer: Maybe you could point toward documentation for these, in case people don't know what they do/why they are important

> - `gapbs` benchmark sutie with ROI annotations

Reviewer: ```suggestion
- `gapbs` benchmark suite with ROI annotations
```

Thus, you should be able to build packages on the disk and easily link to the gem5-bridge library.

The disk has network disabled by default to improve boot time in gem5.

If you want to enable networking, you need to modify the disk image and move the file `/etc/netplan/00-installer-config.yaml.bak` to `/etc/netplan/00-installer-config.yaml`.

> ## Building the Disk Image

Reviewer: Swap the positions of the sections "## Building the Disk Image" and "## What's on the disk?" I think this will make it more convenient for the reader, as it would put the most crucial piece of information on top.

Assuming that you are in the `src/gapbs/` directory, run

```sh
./build.sh          # the script downloading packer binary and building the disk image
```

After this process succeeds, the disk image can be found on the `src/gapbs/disk-image-ubuntu-24-04`.

> This gapbs image uses the prebuilt ubuntu 24.04 image as a base image. The gapbs image also throws the same exit events as the base image.

Reviewer: Elaborate more on exit events in the base disk image and/or add link to base image documentation. 

I think you should also say "Look at the Boot Sequences section for information on what exit events will take place", then tell users where they can find configuration scripts with that exit event setup. Since the PR to update the config scripts is still pending, maybe you could point users to the "usage" tab of one of the gem5 resources with the fleshed out exit event handler setup (such as this one: https://resources.gem5.org/resources/x86-ubuntu-24.04-img/usage?database=gem5-resources&version=2.0.0).  Or even just put it in a code block somewhere in this document.

Reviewer 2: I agree with more explanations about two things:
1. what the path to the programs' executable is in the disk image. This is probably required to setup the command in the run script. I saw in the examples Erin provided below, she has not specified that, and has  used them directly as `./cc` for this purpose.
2. number of exit events, their orders, and where they are in the main program.

Author: 1. I have added the path to the executable, I will also add the links to the workloads on gem5 resources when the changes looks good, and the disks have been uploads. There will be a workload for each executable benchmark in 3 sizes and with two real graphs. Is there any other information that I should add?

2. The boot sequences section in the readme explains, where the exits will occur. will it be beneficial to mention exact files?

Reviewer 2: 1. Thanks for adding this. Looks good to me.
2. That also looks good with the stars marked to see how many exit events must be handled in the runs scripts.

> Each benchmark also has its regions of intrests annotated and they throw a `gem5-bridge workbegin` and `gem5-bridge workend` exit event.

Reviewer: ```suggestion
Each benchmark also has its regions of interest annotated and they throw a `gem5-bridge workbegin` and `gem5-bridge workend` exit event.
```

### Boot Sequences

#### Default Boot Sequence (Systemd, Non-Interactive)

- Kernel output
- **Kernel Booted print message** **
- Running systemd print message
- Systemd output
- autologin
- **Running after_boot script** **
- Print indicating **non-interactive** mode
- **Reading run script file**
- Script output
- Exit **

#### With Systemd and Interactive

- Kernel output
- **Kernel Booted print message** **
- Running systemd print message
- Systemd output
- autologin
- **Running after_boot script** **
- Shell

#### Without Systemd and Non-Interactive

- Kernel output
- **Kernel Booted print message** **
- autologin
- **Running after_boot script** **
- Print indicating **non-interactive** mode
- **Reading run script file**
- Script output
- Exit **

#### Without Systemd and Interactive

- Kernel output
- **Kernel Booted print message** **
- autologin
- **Running after_boot script** **
- Shell

This detailed overview provides a foundational understanding of how different boot configurations affect the system's initialization and mode of operation.

> By selecting the appropriate parameters, users can customize the boot process for diverse environments, ranging from automated setups to hands-on interactive sessions.

Reviewer: It could be good to add an overview of the benchmarks available on the built disk, or point toward the GAPBS documentation

Author: I will add a section with all the benchmark workloads, when we upload the disk to the cloud and I can push the JSONs to MongoDB.

permalink: resources/gapbs
shortdoc: >
    This resource implementes the [GAP benchmark suite](http://gap.cs.berkeley.edu/benchmark.html).
author: ["Harshil Patel"]
license: BSD-3-Clause
---

This document provides instructions to create a GAP Benchmark Suite (GAPBS) disk image, which, along with an example script, may be used to run GAPBS within gem5 simulations. The example script uses a pre-built disk-image.

A pre-built disk image, for X86, can be found, gzipped, here: <add link>

## Building the Disk Image

Assuming that you are in the `src/gapbs/` directory, run

```sh

> ./build.sh          # the script downloading packer binary and building the disk image

Reviewer: ```suggestion
./build-{ISA}.sh          # the script downloading packer binary and building the disk image. {ISA} is `x86`, `arm`, or `riscv`. 
```

- Running systemd print message
- Systemd output
- autologin
- **Running after_boot script** **
- Shell

#### Without Systemd and Non-Interactive

- Kernel output
- **Kernel Booted print message** **
- autologin
- **Running after_boot script** **
- Print indicating **non-interactive** mode
- **Reading run script file**
- Script output
- Exit **

#### Without Systemd and Interactive

- Kernel output
- **Kernel Booted print message** **
- autologin
- **Running after_boot script** **
- Shell

This detailed overview shows how different boot configurations affect the system's initialization and mode of operation.
By selecting the appropriate parameters, users can customize the boot process for diverse environments, ranging from automated setups to hands-on interactive sessions.

## Using the disk image and kernel in gem5

All GAPBS benchmarks are available on <https://resources.gem5.org/>.
You can use these benchmarks directly in your gem5 config scripts, like this:

```python
board.set_workload(
    obtain_resource("arm-ubuntu-24.04-gapbs-bc-test", resource_version="1.0.0")
)
```

If you have made the disk image locally and want to run it on gem5 you can use the following templates:

> **Note**: You would need to use the kernels mentions in the templates to make sure gem5 can run the workload.

Reviewer: ```suggestion
**Note**: You would need to use the kernels mentioned in the templates to make sure gem5 can run the workload.
This is because these disk images are built with the [gem5 bridge driver module](https://github.com/gem5/gem5/tree/stable/util/gem5_bridge), added in [PR 1480](https://github.com/gem5/gem5/pull/1480), which makes it so that gem5 simulations don't need sudo permissions when running commands. The gem5 bridge driver module has the limitation of needing to be run with the same kernel version as what it was built with.
```

Reviewer: The above suggestion should also be split up so each line isn't too long.
